python/pdo/contract/response.py

```diff
@@ -90,6 +100,216 @@ class ContractResponse(object) :
     Class for managing the contract operation response from an enclave service
     """
 
+    __start_threads_for_commit__ = True
+    transaction_submission_exceptions_queue = queue.Queue()
+
+    transaction_executor = concurrent.futures.ThreadPoolExecutor(max_workers=1) # executor that submit transactions
+    pending_transactions_queue = queue.Queue()
+
+    txn_completion_condition = threading.Condition()
+    set_of_completed_txns = set()
+    
+    # -------------------------------------------------------
+    @classmethod
+    def get_exceptions_from_past_commits(cls):
+        """ Return any exception rasied by concurrent Replication and transaction submsission tasks"""
+
+        try:
+            e = Replicator.exceptions_queue.get_nowait()
+        except:
+            try:
+                e = cls.transaction_submission_exceptions_queue.get_nowait()
+            except:
+                e = None
+
+        return e
+
+    # -------------------------------------------------------
+
+    @classmethod
+    def wait_for_commit(cls, commit_id, use_ledger=True, shut_down=False):
+        """ Wait for completion of the commit task specified by id. Return transaction id if ledger is used, else return None. If shut_down is True,
+        send exit singals to replication and transaction threads. The implementation relies on threading.Condition objects"""
+        
+        if use_ledger:
+            cls.txn_completion_condition.acquire()
+            while commit_id not in cls.set_of_completed_txns: 
+                cls.txn_completion_condition.wait(timeout=1.0)
+                e = cls.get_exceptions_from_past_commits()
+                if e is not None: 
+                    raise Exception(str(e))
+                cls.txn_completion_condition.acquire()    
+            cls.txn_completion_condition.release()    
+
+            global transaction_dependencies
+            contract_id = commit_id[0]
+            state_hash = commit_id[1]
+            txn_id = transaction_dependencies.FindDependencyLocally(contract_id, crypto.byte_array_to_base64(state_hash))
+        
+        else: #wait until the replication task for commit_id completes
+
+            Replicator.replication_completion_condition.acquire()
+            while commit_id not in Replicator.set_of_completed_replications: 
+                Replicator.replication_completion_condition.wait(timeout = 1.0)
+                e = cls.get_exceptions_from_past_commits()
+                if e is not None:
+                    raise Exception(str(e))
+                Replicator.replication_completion_condition.acquire()    
+            Replicator.replication_completion_condition.release()    
+
+            txn_id = None
+
+        if shut_down:
+            cls.exit_commit_workers()
+        
+        return txn_id
+
```

Comment:
again... the number of calls to class functions should be an indication that this is not really part of the class but really part of a separate module. if the method is specific to managing this instance of the wait (which is what I would have thought) then it isn't really a class method. So I'm not sure how you see the pieces fitting together.

Reply:
Yes, I agree, please see my answer above, a separate module for transaction/commit processing makes sense. 

And yes, thank you, I agree it makes more sense to make it an instance method than a class method.

Reply:
sorry i take back, it makes more sense to keep the wait method as a class method instead of an isntance method. The reason is that we want to wait on some specific commit_id after say 5 updats, not necessarily the 5th one. There are several reasons: the 5th one failed, 5th did not change, in both these cases, we want to jsut wait for the 4th one to complete. 

If we decide to make the wait as instance function, then we need to carry out the previous repsonse object, where as for the class method, we just need the commi_id for the 4th response.